src/api_authentication_app_skeleton/spec/support/app_visitor.cr

```diff
@@ -0,0 +1,131 @@
+
+  getter! response
+  @response : HTTP::Client::Response? = nil
+
+  def visit(path : String, as_user : Authentic::PasswordAuthenticatable? = nil, headers : HTTP::Headers? = nil)
+    if as_user
+      headers ||= HTTP::Headers.new
+      headers.add "Authorization", generate_token(as_user)
+    end
+
+    request = HTTP::Request.new("GET", path, headers)
+    @response = process_request(request)
+  end
+
+  def put(path : String, body : Hash(String, String), as_user : Authentic::PasswordAuthenticatable? = nil, headers : HTTP::Headers? = nil)
+    if as_user
+      headers ||= HTTP::Headers.new
+      headers.add "Authorization", generate_token(as_user)
+    end
+
+    request_with_body("PUT", path, with_csrf_token(body), headers)
+  end
+
+  def post(path : String, body : Hash(String, String), as_user : Authentic::PasswordAuthenticatable? = nil, headers : HTTP::Headers? = nil)
+    if as_user
+      headers ||= HTTP::Headers.new
+      headers.add "Authorization", generate_token(as_user)
+    end
+
+    request_with_body("POST", path, with_csrf_token(body), headers)
+  end
+
+  def delete(path : String, as_user : Authentic::PasswordAuthenticatable? = nil, headers : HTTP::Headers? = nil)
+    if as_user
+      headers ||= HTTP::Headers.new
+      headers.add "Authorization", generate_token(as_user)
+    end
+
+    request = HTTP::Request.new("DELETE", path, headers)
+    @response = process_request(request)
+  end
+
+  def response_body
+    body = @response.not_nil!.body
+    if body[0, 5] == "ERROR"
+      raise Exception.new(body)
+    elsif body == ""
+      JSON.parse("{}")
+    else
+      JSON.parse(body)
+    end
+  end
+
+  private def with_csrf_token(body : Hash(String, String))
+    body[Lucky::ProtectFromForgery::PARAM_KEY] = CSRF_TOKEN
+    body
+  end
+
+  private def request_with_body(method : String, path : String, body : Hash(String, String), headers : HTTP::Headers? = nil)
+    body_strings = [] of String
+    body.each do |key, value|
+      body_strings << "#{URI.escape(key)}=#{URI.escape(value)}"
+    end
+    request = HTTP::Request.new(method, path, headers, body_strings.join("&"))
+    @response = process_request(request)
+  end
+
+  private def process_request(request)
+    io = IO::Memory.new
+    response = HTTP::Server::Response.new(io)
+    context = HTTP::Server::Context.new(request, response)
+    context.session[Lucky::ProtectFromForgery::SESSION_KEY] = CSRF_TOKEN
+    middlewares.call context
+    response.close
+    io.rewind
+    HTTP::Client::Response.from_io(io, decompress: false)
+  end
+
+  def middlewares
+    HTTP::Server.build_middleware([
+      Lucky::HttpMethodOverrideHandler.new,
+      Lucky::LogHandler.new,
+      Lucky::SessionHandler.new,
+      Lucky::Flash::Handler.new,
+      Lucky::ErrorHandler.new(action: Errors::Show),
+      Lucky::RouteHandler.new,
+      Lucky::StaticFileHandler.new("./public", false),
+      Lucky::RouteNotFoundHandler.new,
+    ])
+  end
```

Comment:
I've been thinking about this and I think it is a bit brittle. It relies on copying the middleware and doesn't fully test how a web request comes through the server. I think instead it would be better to use [`HTTP::Client`](https://crystal-lang.org/api/0.25.1/HTTP/Client.html) to make a real request against the server. 

You can use Lucky::RouteHelper.settings.base_uri to get the base_uri for the request.

For this to work you'll also need to move: https://github.com/luckyframework/lucky_cli/blob/master/src/browser_app_skeleton/spec/setup/start_app_server.cr to `web_app_base_skeleton` so it is available to browser and api apps. It will need to conditionally shutdown LuckyFlow for just browser apps. Does this make sense?

It's kind of complicated so LMK if I should explain some more

Reply:
Yes I see what you mean. I'll try to set it up